pkg/container-collection/container-collection.go

```diff
@@ -0,0 +1,228 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package containercollection provides the ContainerCollection struct to keep
+// track of the set of running containers and primitives to query that set with
+// various criteria.
+//
+// It is used by the Gadget Tracer Manager to keep track of containers part of
+// Kubernetes pods and by Local Gadget Manager to keep track of containers on a
+// Linux system.
+package containercollection
+
+import (
+	"sync"
+
+	pb "github.com/kinvolk/inspektor-gadget/pkg/gadgettracermanager/api"
+	"github.com/kinvolk/inspektor-gadget/pkg/gadgettracermanager/pubsub"
+)
+
+// ContainerCollection holds a set of containers. It can be embedded as an
+// anonymous struct to help other structs implement the ContainerResolver
+// interface. For this reason, some methods are prefixed with 'Container' to
+// make clear that
+
+type ContainerCollection struct {
+	// Keys:   Container ID
+	// Values: pb.ContainerDefinition
+	containers sync.Map
+
+	// subs contains a list of subscribers of container events
+	pubsub *pubsub.GadgetPubSub
+}
+
+// ContainerCollectionInitialize initializes a ContainerCollection. It is
+// useful when ContainerCollection is embedded as an anonymous struct because
+// we don't use a contructor in that case.
+func (cc *ContainerCollection) ContainerCollectionInitialize() {
+	cc.pubsub = pubsub.NewGadgetPubSub()
+}
+
+// ContainerGet looks up a container by the container id and return it if
+// found. The ok result indicates whether value was found in the map.
+func (cc *ContainerCollection) ContainerGet(id string) (value *pb.ContainerDefinition, ok bool) {
+	v, ok := cc.containers.Load(id)
+	if !ok {
+		return nil, false
+	}
+	containerDefinition := v.(pb.ContainerDefinition)
+	return &containerDefinition, true
+}
+
+// ContainerRemove removes a container from the collection.
+func (cc *ContainerCollection) ContainerRemove(id string) {
+	c, ok := cc.ContainerGet(id)
+	if !ok {
+		return
+	}
+	cc.pubsub.Publish(pubsub.EVENT_TYPE_REMOVE_CONTAINER, *c)
+	cc.containers.Delete(id)
+}
+
+// ContainerAdd adds a container to the collection.
+func (cc *ContainerCollection) ContainerAdd(container pb.ContainerDefinition) {
+	cc.containers.Store(container.Id, container)
+	cc.pubsub.Publish(pubsub.EVENT_TYPE_ADD_CONTAINER, container)
+}
+
+// LookupMntnsByContainer returns the mount namespace inode of the container
+// specified in arguments or zero if not found
+func (cc *ContainerCollection) LookupMntnsByContainer(namespace, pod, container string) (mntns uint64) {
+	cc.containers.Range(func(key, value interface{}) bool {
+		c := value.(pb.ContainerDefinition)
+		if namespace != c.Namespace {
+			return true
+		}
+		if pod != c.Podname {
+			return true
+		}
+		if container != c.Name {
+			return true
+		}
+		mntns = c.Mntns
```

Comment:
Can we take the opportunity to rewrite it as?

```
if namespace == c.Namespace &6 pod == c.Podname && container == c.Name {
 ...
}
```

I think it's just cleaner and easier to understand. Probably @marga-kinvolk has comments on this one too. 

My comment applies to other similar instances as well.

Reply:
Done